### PR TITLE
Add support for custom context

### LINKED ISSUE #604: Add custom context
Add these API methods:

* `context.addCustomContext(String key, String value)`
* `context.addCustomContext(String key, Number value)`
* `context.addCustomContext(String key, boolean value)`

See https://github.com/elastic/apm/issues/7#issuecomment-484057924

### PR DESCRIPTION
**Pull request by felixbarny**
closes elastic/apm-agent-java#604

### CODE REVIEW

File: apm-agent-api/src/main/java/co/elastic/apm/api/Transaction.java [hunk 1]
```diff
@@ -79,6 +79,48 @@
...
+     * Non-indexed means the data is not searchable or aggregatable in Elasticsearch,
+     * and you cannot build dashboards on top of the data.
+     * However, non-indexed information is useful for other reasons,
+     * like providing contextual information to help you quickly debug performance issues or errors.
+     *
+     * @param key   The custom context key.
+     * @param value The custom context value.
+     */
+    @Nonnull
+    Transaction addCustomContext(String key, String value);
```
Comment (felixbarny) [inline apm-agent-api/src/main/java/co/elastic/apm/api/Transaction.java]: An alternative would be to add a `getContext` method and add the `addCustom()` methods to the `TransactionContext`. Also kinda depends on how we want to make the other context objects accessible. Also, as of 7.x there is no `context` namespace in ES documents anymore. Context only exists in the intake API now. Custom is stored under `transaction.custom`. So maybe we should just name the method `addCustom`?

Comment (eyalkoren) [reply apm-agent-api/src/main/java/co/elastic/apm/api/Transaction.java]: I think it is good as it is. The term `context` is appropriate here regardless of the schema we use.
With `getContext` do you mean we add an API context struct for data storage? Not sure this is necessary.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/DslJsonSerializer.java [hunk 2]
```diff
@@ -715,6 +715,11 @@ private void serializeContext(final TransactionContext context, TraceContext tra
         }
         serializeRequest(context.getRequest());
         serializeResponse(context.getResponse());
+        if (context.hasCustom()) {
+            writeFieldName("custom");
+            serializeLabels(context.getCustomIterator(), replaceBuilder, jw);
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/DslJsonSerializer.java]: Maybe change the method name to `serializeDictionary` or something of that sort instead?

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/TransactionContext.java [hunk 3]
```diff
@@ -55,17 +59,34 @@ public void copyFrom(TransactionContext other) {
...
+    public Object getCustom(String key) {
+        return custom.get(key);
     }
 
     public Response getResponse() {
         return response;
     }
 
+    public void addCustom(String key, String value) {
+        custom.put(key, value);
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/TransactionContext.java]: No size limitation at all on those?

File: apm-agent-api/src/main/java/co/elastic/apm/api/TransactionImpl.java [hunk 4]
```diff
@@ -77,6 +77,24 @@ public Transaction addLabel(String key, boolean value) {
         return this;
     }
 
+    @Nonnull
+    @Override
+    public Transaction addCustomContext(String key, String value) {
+        return this;
```
Comment (eyalkoren) [inline apm-agent-api/src/main/java/co/elastic/apm/api/TransactionImpl.java]: What about your nice comments referring to the corresponding instrumentation? I like those 😄

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/TransactionContext.java (hunk 3 above)
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/TransactionContext.java]: the size limitation is enforced centrally on serialization

Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/TransactionContext.java]: Ahh. Don't we want to allow a bigger size for these? Probably a common use for non-indexed text is big texts that users still want to keep, like HTTP bodies.

Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/TransactionContext.java]: Oh, good point. We'll want to use the 10k char limit on those values!

**Comment by felixbarny** [comment]
> I assume there was a discussion about that and this is aligned with other agents, I just don't recall- what is the benefit of custom Numbers and Boolean? Would it make a big difference if indexed (as labels)?

Do you mean why we can't simply convert numbers and booleans as strings and store them as keywords?

It's because you can't graph strings if you want the actual number to be on the y-axis. In other words, you can't use [metric aggregations](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics.html) on strings.

File: apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/TransactionInstrumentation.java [hunk 5]
```diff
@@ -101,4 +102,25 @@ private static void ensureParentId(@Advice.FieldValue(value = "span", typing = A
...
     }
+
+    public static class AddCustomContextInstrumentation extends TransactionInstrumentation {
+        public AddCustomContextInstrumentation() {
+            super(named("addCustomContext"));
+        }
+
+        @VisibleForAdvice
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        public static void addLabel(@Advice.FieldValue(value = "span", typing = Assigner.Typing.DYNAMIC) TraceContextHolder<?> context,
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/TransactionInstrumentation.java]:
```suggestion
        public static void addCustomContext(@Advice.FieldValue(value = "span", typing = Assigner.Typing.DYNAMIC) TraceContextHolder<?> context,
```

Comment (felixbarny) [reply apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/TransactionInstrumentation.java]: oops, a classical copy/pasta 🍝

**Pull request merged**